extendr-api/src/iter.rs

```diff
@@ -178,34 +186,61 @@ pub trait AsStrIter: GetSexp + Types + Length + Attributes + Rinternals {
     fn as_str_iter(&self) -> Option<StrIter> {
         let i = 0;
         let len = self.len();
-        match self.sexptype() {
-            SEXPTYPE::STRSXP => unsafe {
+        if self.sexptype() == SEXPTYPE::STRSXP {
```

Comment:
I would say this implementation (though it already was written like this is flawed). We use the same iterator to process three different types only to 'differentiate' between them at the element access level. It should be something like `Option<Box<dyn StrIter>>`. But maybe not for this PR

Reply:
Yes.. You're right. So what I thought was.. the factor part should be its own thing anyways, and started a draft for that here #753. But you're right. This should happen at the instance level, and maybe be split by enum or something like that. I'll give it a go in another PR, for now, let us clean this up and take the win from the fast conversion of `CHARSXP` afforded.

I'd appreciate an issue on it.